tokio-util/src/codec/mod.rs

```diff
@@ -1,18 +1,255 @@
+//!
+//!         // Check that the length is not too large to avoid a denial of
+//!         // service attack where the server runs out of memory.
+//!         if length > 8 * 1024 * 1024 {
+//!             return Err(std::io::Error::new(
+//!                 std::io::ErrorKind::InvalidData,
+//!                 format!("Frame of length {} is too large.", length)
+//!             ));
+//!         }
+//!
+//!         if src.len() < 4 + length {
+//!             // The full string has not yet arrived.
+//!             //
+//!             // We reserve more space in the buffer. This is not strictly
+//!             // necessary, but is a good idea performance-wise.
+//!             src.reserve(4 + length - src.len());
+//!
+//!             // We inform the Framed that we need more bytes to form the next
+//!             // frame.
+//!             return Ok(None);
+//!         }
+//!
+//!         // Use advance to modify src such that it no longer contains
+//!         // this frame.
+//!         let data = src[4..4 + length].to_vec();
+//!         src.advance(4 + length);
+//!
+//!         // Convert the data to a string, or fail if it is not valid utf-8.
+//!         match String::from_utf8(data) {
+//!             Ok(string) => Ok(Some(string)),
+//!             Err(utf8_error) => {
+//!                 Err(std::io::Error::new(
+//!                     std::io::ErrorKind::InvalidData,
+//!                     utf8_error.utf8_error(),
+//!                 ))
+//!             },
+//!         }
+//!     }
+//! }
+//! ```
+//!
+//! # The Encoder trait
+//!
+//! An [`Encoder`] is used together with [`FramedWrite`] or [`Framed`] to turn
+//! an [`AsyncWrite`] into a [`Sink`]. The job of the encoder trait is to
+//! specify how frames are turned into a sequences of bytes.  The job of the
+//! `FramedWrite` is to take the resulting sequence of bytes and write it to the
+//! IO resource.
+//!
+//! The main method on the `Encoder` trait is the [`encode`] method. This method
+//! takes an item that is being written, and a buffer to write the item to. The
+//! buffer may already contain data, and in this case, the encoder should append
+//! the new frame the to buffer rather than overwrite the existing data.
+//!
+//! It is guaranteed that, from one call to `encode` to another, the provided
+//! buffer will contain the exact same data as before, except that some of the
+//! data may have been removed from the front of the buffer. Writing to a
+//! `FramedWrite` is essentially equivalent to the following loop:
+//! ```no_run
+//! use tokio::io::AsyncWriteExt;
+//! use bytes::Buf; // for advance
+//! # use tokio_util::codec::Encoder;
+//! # async fn next_frame() -> bytes::Bytes { bytes::Bytes::new() }
+//! # async fn no_more_frames() { }
+//! # #[tokio::main] async fn main() -> std::io::Result<()> {
+//! # let mut io_resource = tokio::io::sink();
+//! # let mut encoder = tokio_util::codec::BytesCodec::new();
+//! # const MAX: usize = 8192;
```

Comment:
I would consider that we expose some of these items in the doc instead of hiding them?